src/components/Popup.tsx

```diff
@@ -0,0 +1,200 @@
+import React, { useEffect, useRef, useState } from 'react'
+import { useNavigate } from 'react-router-dom';
+import visa from './../assets/icons/Visa_Inc._logo.png'
+import master from './../assets/icons/Mastercard-logo.png'
+import unionPay from './../assets/icons/UnionPay_logo.png'
+
+const Popup: React.FunctionComponent<{
+    reverse: ()=>boolean;
+    hidden : boolean}> = (props) => {
+    const history = useNavigate();
+    const [nameDirty, setNameDirty] = useState(-1)
+    const [numberDirty, setNumberDirty] = useState(-1)
+    const [addressDirty, setAddressDirty] = useState(-1)
+    const [emailDirty, setEmailDirty] = useState(-1)
+    const [cardDirty, setCardDirty] = useState(-1)
+    const [validDirty, setValidDirty] = useState(-1)
+    const [codeDirty, setCodeDirty] = useState(-1);
+    const [cvv, setCvv] = useState(300);
+    const [hidden, setHidden] = useState(false);
+    const [hidepop, setHidepop] = useState(false);
+
+
+    const isAllValid = () : boolean =>{
+        let isValid: number = 1;
+        if(!!nameDirty){
+            setNameDirty(1);
+            isValid = 0
+        }
+        if(!!numberDirty){
+            setNumberDirty(1);
+            isValid = 0
+        }
+        if(!!addressDirty){
+            setAddressDirty(1);
+            isValid = 0
+        }
+        if(!!emailDirty){
+            setEmailDirty(1);
+            isValid = 0
+        }
+        if(!!cardDirty){
+            setCardDirty(1);
+            isValid = 0
+        }
+        if(!!validDirty){
+            setValidDirty(1);
+            isValid = 0
+        }
+        if(!!codeDirty){
+            setCodeDirty(1);
+            isValid = 0
+        }
+        return !!isValid
+    }
+
+    function confrimForm(){
+        if(isAllValid()){
+            setHidepop(true)
+            localStorage.setItem("cart",'[]')
+            setTimeout(() => {
+                history("/")
+            }, 4000);
+            
+        }
+    }
+
+    let lastVal: string = ""
+    const setSlash = (e: {target: HTMLInputElement}):void => {
+        if(lastVal.length < e.target.value.length && e.target.value.length === 2) {
```

Comment:
It is difficult to read. You can create `const isSlashNeeded = lastVal.length < e.target.value.length && e.target.value.length === 2` and then use it as `if (isSlashNeeded) {}`